Application/ControlMsrE2/ControlMsrE2.c

```diff
@@ -0,0 +1,252 @@
+      Print (L"No corresponding BIOS Options found. Try a different search string using interactive mode.\n");
+    } else {
+      Key = '1';
+
+      if (OptionsCount > 1) {
+        do {
+          Print (L"\nEnter choice (1..%x) ? ", OptionsCount);
+          Key = ReadAnyKey ();
+        } while ((Key < '1') && (Key > '0' + OptionsCount) && (Key != 0x1B));
+
+        Print (L"\n");
+      }
+
+      if (Key != 0x1B) {
+        Index = Key - '0';
+
+        for (ContextIndex = 0; ContextIndex < ContextsCount; ++ContextIndex) {
+          if (Contexts[ContextIndex].Count >= Index) {
+            Contexts[ContextIndex].Count = ContextIndex == 0 ? 0 : Contexts[ContextIndex - 1].Count;
+            Contexts[ContextIndex].StopAt = Index;
+            Contexts[ContextIndex].IfrOneOf = NULL;
+
+            IterateOpCode (
+              Contexts[ContextIndex].FirstIfrHeader,
+              EFI_IFR_ONE_OF_OP,
+              &Stop,
+              &Contexts[ContextIndex],
+              HandleIfrOption
+              );
+
+            if (Contexts[ContextIndex].IfrOneOf != NULL) {
+              HandleIfrVariable (&Contexts[ContextIndex]);
+            }
+            break;
+          }
+        }
+      }
+    }
+  }
+
+  for (ListHeaderIndex = 0; ListHeaderIndex < ListHeaderCount; ++ListHeaderIndex) {
+    if (ListHeaders[ListHeaderIndex] != NULL) {
+      FreePool (ListHeaders[ListHeaderIndex]);
+    }
+  }
+}
+
+EFI_STATUS
+SearchForString (
+  IN EFI_STRING SearchString
+  )
+{
+  EFI_HII_HANDLE                *HiiHandles;
+  EFI_HII_PACKAGE_LIST_HEADER   **ListHeaders;
+  UINT32                        ListHeaderCount;
+
+  if (IS_INTERACTIVE ()) {
+    ModifySearchString (&SearchString);
+  }
+
+  HiiHandles = HiiGetHiiHandles (NULL);
+  if (HiiHandles == NULL) {
+    Print (L"Could not retrieve HiiHandles.\n");
+    return EFI_OUT_OF_RESOURCES;
+  }
+
+  //
+  // Retrieve ListHeaderCount
+  //
+  for (ListHeaderCount = 0; HiiHandles[ListHeaderCount] != NULL; ++ListHeaderCount);
+
+  //
+  // Keep list alive 'til program finishes.
+  // So that all lists can be searched, the results be displayed together.
+  // And from all those one Option will be selected to be changed
+  //
+  ListHeaders = AllocatePool (sizeof (*ListHeaders) * ListHeaderCount);
+  if (ListHeaders == NULL) {
+    Print (L"Could not allocate memory for ListHeaders.\n");
+    return EFI_OUT_OF_RESOURCES;
+  }
+
+  IterateListHeaders (HiiHandles, ListHeaders, ListHeaderCount, SearchString);
+  FreePool (ListHeaders);
+  return EFI_SUCCESS;
+}
+
+EFI_STATUS
+EFIAPI
+UefiMain (
+  IN EFI_HANDLE        ImageHandle,
+  IN EFI_SYSTEM_TABLE  *SystemTable
+  )
+{
+  EFI_STATUS Status;
+  EFI_STRING SearchString;
+
+  Status = InterpretArguments ();
+  if (!EFI_ERROR (Status)) {
+    Status = VerifyMSRE2 ();
+    if (!EFI_ERROR (Status)) {
+      if (Flags != ARG_VERIFY) {
+        Print (L"\nBIOS Options:\n");
+
+        SearchString = AsciiStrCopyToUnicode ("cfg", 0);
```

Comment:
Why would you call this function with a literal? In my opinion: `AllocateCopyPool (L_STR_SIZE(L"cfg"), L"cfg");`